src/coreclr/jit/codegenlinear.cpp

```diff
@@ -712,7 +712,9 @@ void CodeGen::genCodeForBBlist()
 
                     if ((call != nullptr) && (call->gtOper == GT_CALL))
                     {
-                        if ((call->AsCall()->gtCallMoreFlags & GTF_CALL_M_DOES_NOT_RETURN) != 0)
+                        if ((call->AsCall()->gtCallMoreFlags & GTF_CALL_M_DOES_NOT_RETURN) != 0 ||
```

Comment:
Can we extend `GenTreeCall::IsNoReturn()` to cover the helper case too? Either by setting the  no return flag when the helper is set, or by putting this check there?

Reply:
I have tried, naively, to put 

```c
    if (s_helperCallProperties.AlwaysThrow((CorInfoHelpFunc)helper))
    {
        result->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
    }
```

in gtNewHelperCallNode

, but that resulted in 

```
  Assertion failed 'call->gtCallType == CT_USER_FUNC' in 'System.ModuleHandle:ResolveFieldHandle(int,System.RuntimeTypeHandle[],System.Run
  timeTypeHandle[]):System.RuntimeFieldHandle:this' during 'Merge throw blocks' (IL size 283; hash 0x040825d3; FullOpts)
```

```c
        if (!call->IsNoReturn())
        {
            continue;
        }

        // Sanity check -- only user funcs should be marked do not return
        assert(call->gtCallType == CT_USER_FUNC);
```

I could not easily tell if the assert is just an observation/sanity, that could be relaxed or indeed guarding some assumptions made in the code, so I did not continue on that path.

Reply:
Commenting out that assert leads to others:

```
  Assert failure(PID 35472 [0x00008a90], Thread: 41468 [0xa1fc]): Assertion failed 'updateCount < optNoReturnCallCount' in 'System.Tests.S
  tringComparerTests:Create_CreatesValidComparer():this' during 'Merge throw blocks' (IL size 543; hash 0xc45bc068; FullOpts)
```

Reply:
I think I'd prefer entering an issue on this.

Reply:
Makes sense, looks like there are some assumptions in the throw helper merging to sort out.